File: lib/consts.js
// Source-engine movement constants. Single source of truth — every other module reads
// from here rather than hardcoding, so there is exactly one place to correct if a value
// turns out to be wrong.
//
// CS:GO / Source 1 values. CS2 differs and is explicitly out of scope.
"use strict";

// ---------------------------------------------------------------- physics
const GRAVITY = 800;                        // sv_gravity
const JUMP_IMPULSE = Math.sqrt(2 * GRAVITY * 57);  // sv_jump_impulse ~= 301.993377
const STEP_SIZE = 18;                       // sv_stepsize — free height gain while walking

// ---------------------------------------------------------------- player hull
// The standing hull is 32x32x72; ducking shrinks it to 32x32x54. Ducking in mid-air keeps
// the hull centred, so the feet rise by half the 18u shrink — this is exactly why a crouch
// jump reaches 9.00u higher than a normal one (see jumptable.js).
const HULL_W = 32;
const HULL_H_STAND = 72;
const HULL_H_DUCK = 54;
const DUCK_FEET_GAIN = (HULL_H_STAND - HULL_H_DUCK) / 2;   // 9.00

// Eye offsets (view_ofs). These are the numbers cl_showpos reports, so they are what the
// user reads off their screen when lining a surf up.
const EYE_STAND = 64.09;
const EYE_DUCK = 46.07;

// Hitbox heights used when stacking players for a boost. The engine reports a hair less
// for players above the first in a stack, which is why 72.04 and 72.03 both exist.
const BOX_STAND = 72.04;
const BOX_STAND_UPPER = 72.03;
const BOX_DUCK = 54.04;
const BOX_DUCK_UPPER = 54.03;

// ---------------------------------------------------------------- surfaces
// PM_CategorizePosition: a plane counts as ground only if its normal points up enough.
// Anything shallower is a surf ramp — you slide along it instead of standing on it.
const STANDABLE_NORMAL_Z = 0.7;

// ---------------------------------------------------------------- speeds
const SPEED_WALK = 250;      // +speed held
const SPEED_RUN = 320;       // knife/pistol; rifles are slower, see WEAPON_SPEED
const SPEED_MAX_GROUND = 320;
const AIR_SPEED_CAP = 30;    // the 30 u/s air-accel window that makes strafing work
const AIR_ACCELERATE = 12;   // sv_airaccelerate
const ACCELERATE = 5.5;      // sv_accelerate
const FRICTION = 5.2;        // sv_friction
const STOP_SPEED = 80;       // sv_stopspeed

// ---------------------------------------------------------------- bsp contents flags
const CONTENTS_SOLID = 0x1;
const CONTENTS_WATER = 0x20;
const CONTENTS_PLAYERCLIP = 0x10000;
const CONTENTS_MONSTERCLIP = 0x20000;
const CONTENTS_LADDER = 0x20000000;

// Everything that stops a player. Grate/window contents are deliberately excluded — they
// block bullets, not movement.
const CONTENTS_PLAYER_SOLID = CONTENTS_SOLID | CONTENTS_PLAYERCLIP;

// ---------------------------------------------------------------- tickrates
const TICKRATES = [64, 128];

module.exports = {
  GRAVITY, JUMP_IMPULSE, STEP_SIZE,
  HULL_W, HULL_H_STAND, HULL_H_DUCK, DUCK_FEET_GAIN,
  EYE_STAND, EYE_DUCK,
  BOX_STAND, BOX_STAND_UPPER, BOX_DUCK, BOX_DUCK_UPPER,
  STANDABLE_NORMAL_Z,
  SPEED_WALK, SPEED_RUN, SPEED_MAX_GROUND, AIR_SPEED_CAP, AIR_ACCELERATE,
  ACCELERATE, FRICTION, STOP_SPEED,
  CONTENTS_SOLID, CONTENTS_WATER, CONTENTS_PLAYERCLIP, CONTENTS_MONSTERCLIP,
  CONTENTS_LADDER, CONTENTS_PLAYER_SOLID,
  TICKRATES,
};

File: core/src/consts.rs
//! Source-engine movement constants — CS:GO / Source 1. CS2 differs and is out of scope.
//!
//! Single source of truth: nothing else in the crate hardcodes these, so there is exactly
//! one place to correct if a value turns out to be wrong.

// ---------------------------------------------------------------- physics
pub const GRAVITY: f64 = 800.0;                 // sv_gravity
pub const JUMP_APEX: f64 = 57.0;                // a standing jump peaks exactly here
/// sv_jump_impulse ~= 301.993377, i.e. sqrt(2 * g * 57).
pub fn jump_impulse() -> f64 { (2.0 * GRAVITY * JUMP_APEX).sqrt() }
pub const STEP_SIZE: f64 = 18.0;                // sv_stepsize — free height while walking

// ---------------------------------------------------------------- player hull
pub const HULL_W: f64 = 32.0;
pub const HULL_H_STAND: f64 = 72.0;
pub const HULL_H_DUCK: f64 = 54.0;
/// Ducking mid-air keeps the hull centred, so the feet rise by half the shrink. This is
/// exactly why a crouch jump reaches 9.00u higher than a normal one.
pub const DUCK_FEET_GAIN: f64 = (HULL_H_STAND - HULL_H_DUCK) / 2.0;   // 9.00

/// Eye offsets (view_ofs) — what cl_showpos reports, so what the user reads off screen.
pub const EYE_STAND: f64 = 64.09;
pub const EYE_DUCK: f64 = 46.07;

/// Hitbox heights for boost stacks. The engine reports a hair less for players above the
/// first in a stack, which is why both 72.04 and 72.03 exist.
pub const BOX_STAND: f64 = 72.04;
pub const BOX_STAND_UPPER: f64 = 72.03;
pub const BOX_DUCK: f64 = 54.04;
pub const BOX_DUCK_UPPER: f64 = 54.03;

// ---------------------------------------------------------------- surfaces
/// PM_CategorizePosition: a plane is ground only if its normal points up at least this much.
/// 0.7 is 45.573 degrees — so an exactly-45-degree ramp is still a floor, and surf ramps are
/// the ones steeper than that.
pub const STANDABLE_NORMAL_Z: f64 = 0.7;

// ---------------------------------------------------------------- speeds
pub const SPEED_WALK: f64 = 250.0;
pub const SPEED_RUN: f64 = 320.0;
pub const AIR_SPEED_CAP: f64 = 30.0;    // the window that makes air-strafing work
pub const AIR_ACCELERATE: f64 = 12.0;   // sv_airaccelerate
pub const ACCELERATE: f64 = 5.5;        // sv_accelerate
pub const FRICTION: f64 = 5.2;          // sv_friction
pub const STOP_SPEED: f64 = 80.0;       // sv_stopspeed

// ---------------------------------------------------------------- bsp contents
pub const CONTENTS_SOLID: i32 = 0x1;
pub const CONTENTS_WATER: i32 = 0x20;
pub const CONTENTS_PLAYERCLIP: i32 = 0x10000;
pub const CONTENTS_MONSTERCLIP: i32 = 0x20000;
pub const CONTENTS_LADDER: i32 = 0x2000_0000;

/// Everything that stops a player. Grate/window contents are deliberately excluded — they
/// block bullets, not movement.
pub const CONTENTS_PLAYER_SOLID: i32 = CONTENTS_SOLID | CONTENTS_PLAYERCLIP;

pub const TICKRATES: [u32; 2] = [64, 128];
